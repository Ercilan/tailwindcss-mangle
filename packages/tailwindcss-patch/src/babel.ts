import generate from '@babel/generator'
import { parse } from '@babel/parser'
import traverse from '@babel/traverse'

export { generate, parse, traverse }
