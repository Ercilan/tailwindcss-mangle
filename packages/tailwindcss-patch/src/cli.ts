import type { LegacyTailwindcssPatcherOptions } from './options/legacy'
import type { TailwindcssPatchOptions } from './types'
import process from 'node:process'

import { CONFIG_NAME, getConfig, initConfig } from '@tailwindcss-mangle/config'
import { defu } from '@tailwindcss-mangle/shared'
import cac from 'cac'

import { TailwindcssPatcher } from './api/tailwindcss-patcher'
import logger from './logger'
import { fromLegacyOptions, fromUnifiedConfig } from './options/legacy'

const cli = cac('tw-patch')

async function loadPatchOptions(cwd: string, overrides?: TailwindcssPatchOptions) {
  const { config } = await getConfig(cwd)
  const legacyConfig = config as (typeof config) & { patch?: LegacyTailwindcssPatcherOptions['patch'] }
  const base = config?.registry
    ? fromUnifiedConfig(config.registry)
    : legacyConfig?.patch
      ? fromLegacyOptions({ patch: legacyConfig.patch })
      : {}
  const merged = defu<TailwindcssPatchOptions, TailwindcssPatchOptions[]>(overrides ?? {}, base)
  return merged
}

cli
  .command('install', 'Apply Tailwind CSS runtime patches')
  .option('--cwd <dir>', 'Working directory', { default: process.cwd() })
  .action(async (args: { cwd: string }) => {
    const options = await loadPatchOptions(args.cwd)
    const patcher = new TailwindcssPatcher(options)
    await patcher.patch()
    logger.success('Tailwind CSS runtime patched successfully.')
  })

cli
  .command('extract', 'Collect generated class names into a cache file')
  .option('--cwd <dir>', 'Working directory', { default: process.cwd() })
  .option('--output <file>', 'Override output file path')
  .option('--format <format>', 'Output format (json|lines)')
  .option('--css <file>', 'Tailwind CSS entry CSS when using v4')
  .option('--no-write', 'Skip writing to disk')
  .action(async (args: { cwd: string, output?: string, format?: 'json' | 'lines', css?: string, write?: boolean }) => {
    const overrides: TailwindcssPatchOptions = {}

    if (args.output || args.format) {
      overrides.output = {
        file: args.output,
        format: args.format,
      }
    }

    if (args.css) {
      overrides.tailwind = {
        v4: {
          cssEntries: [args.css],
        },
      }
    }

    const options = await loadPatchOptions(args.cwd, overrides)
    const patcher = new TailwindcssPatcher(options)
    const result = await patcher.extract({ write: args.write })

    if (result.filename) {
      logger.success(`Collected ${result.classList.length} classes → ${result.filename}`)
    }
    else {
      logger.success(`Collected ${result.classList.length} classes.`)
    }
  })

cli
  .command('init', 'Generate a tailwindcss-patch config file')
  .option('--cwd <dir>', 'Working directory', { default: process.cwd() })
  .action(async (args: { cwd: string }) => {
    await initConfig(args.cwd)
    logger.success(`✨ ${CONFIG_NAME}.config.ts initialized!`)
  })

cli.help()
cli.parse()
