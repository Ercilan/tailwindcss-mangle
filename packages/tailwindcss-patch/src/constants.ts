export const pkgName = 'tailwindcss-patch'
