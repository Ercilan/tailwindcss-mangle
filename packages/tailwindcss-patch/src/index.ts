export * from './class'
export * from './exposeContext'
export * from './inspector'
export * from './patcher'
export * from './utils'
export * from './type'
