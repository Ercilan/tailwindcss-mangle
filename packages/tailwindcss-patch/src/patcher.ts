import path from 'path'
import fs from 'fs'
import { gte as semverGte } from 'semver'

import type { PackageJson } from 'pkg-types'

import { inspectPostcssPlugin, inspectProcessTailwindFeaturesReturnContext } from './inspector'

function noop() {}

export function getInstalledPkgJsonPath(options: any) {
  try {
    // const cwd = process.cwd()
    const tmpJsonPath = require.resolve(`tailwindcss/package.json`, {
      paths: options.paths
    })

    const pkgJson = require(tmpJsonPath) as PackageJson
    // https://github.com/sonofmagic/weapp-tailwindcss-webpack-plugin

    if (semverGte(pkgJson.version!, '3.0.0')) {
      return tmpJsonPath
    }
  } catch (error) {
    if ((<Error & { code: string }>error).code === 'MODULE_NOT_FOUND') {
      console.warn('没有找到`tailwindcss`包，请确认是否安装。')
    }
  }
}

export function createPatch(options: any) {
  if (options === false) {
    return noop
  }
  return () => {
    try {
      return internalPatch(getInstalledPkgJsonPath(options), options)
    } catch (error) {
      console.warn(`patch tailwindcss failed:` + (<Error>error).message)
    }
  }
}

function ensureFileContent(filepaths: string | string[]) {
  if (typeof filepaths === 'string') {
    filepaths = [filepaths]
  }
  let content
  for (let i = 0; i < filepaths.length; i++) {
    const filepath = filepaths[i]
    if (fs.existsSync(filepath)) {
      content = fs.readFileSync(filepath, {
        encoding: 'utf-8'
      })
      break
    }
  }
  return content
}
export function monkeyPatchForExposingContext(rootDir: string, overwrite: boolean) {
  const processTailwindFeaturesFilePath = path.resolve(rootDir, 'lib/processTailwindFeatures.js')

  const processTailwindFeaturesContent = ensureFileContent(processTailwindFeaturesFilePath)
  const result: { processTailwindFeatures?: string; plugin?: string } = {}
  if (processTailwindFeaturesContent) {
    const { code, hasPatched } = inspectProcessTailwindFeaturesReturnContext(processTailwindFeaturesContent)
    if (!hasPatched && overwrite) {
      fs.writeFileSync(processTailwindFeaturesFilePath, code, {
        encoding: 'utf-8'
      })
      console.log('patch tailwindcss processTailwindFeatures for return content successfully!')
    }
    result.processTailwindFeatures = code
  }

  const pluginFilePath = path.resolve(rootDir, 'lib/plugin.js')
  const indexFilePath = path.resolve(rootDir, 'lib/index.js')
  const pluginContent = ensureFileContent([pluginFilePath, indexFilePath])
  if (pluginContent) {
    const { code: code0, hasPatched: hasPatched0 } = inspectPostcssPlugin(pluginContent)
    if (!hasPatched0 && overwrite) {
      fs.writeFileSync(pluginFilePath, code0, {
        encoding: 'utf-8'
      })
      console.log('patch tailwindcss for expose runtime content successfully!')
    }
    result.plugin = code0
  }
  return result
}

export function internalPatch(pkgJsonPath: string | undefined, options: any, overwrite: boolean = true): any | undefined {
  if (pkgJsonPath) {
    const rootDir = path.dirname(pkgJsonPath)

    const result = monkeyPatchForExposingContext(rootDir, overwrite)
    return result
  }
}
