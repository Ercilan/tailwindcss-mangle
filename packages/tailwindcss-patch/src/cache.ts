import fs from 'fs'
import path from 'path'
import { pkgName } from './constants'
import type { CacheOptions } from './type'

export function mkCacheDirectory(cacheDirectory: string) {
  const exists = fs.existsSync(cacheDirectory)
  if (!exists) {
    fs.mkdirSync(cacheDirectory, {
      recursive: true
    })
  }
  return cacheDirectory
}

export function getCacheOptions(options: CacheOptions = {}): Required<CacheOptions> & { filename: string } {
  const cwd = options.cwd ?? process.cwd()
  const dir = options.dir ?? path.resolve(cwd, 'node_modules', '.cache', pkgName)
  const file = options.file ?? 'classSet.json'
  const filename = path.resolve(dir, file)
  return {
    cwd,
    dir,
    file,
    filename
  }
}

export function writeCache(data: Set<string>, options: CacheOptions = {}) {
  try {
    const { dir, filename } = getCacheOptions(options)
    mkCacheDirectory(dir)
    fs.writeFileSync(filename, JSON.stringify(Array.from(data), null, 2), 'utf-8')
    return filename
  } catch (error) {
    console.log(error)
  }
}

export function readCache(options: CacheOptions = {}) {
  const { filename } = getCacheOptions(options)
  try {
    const data = fs.readFileSync(filename, 'utf-8')
    return new Set<string>(JSON.parse(data))
  } catch (error) {
    console.log(error)
    fs.unlinkSync(filename)
  }
}
