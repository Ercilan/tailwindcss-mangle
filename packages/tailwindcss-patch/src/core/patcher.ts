import type { PatchUserConfig } from '@tailwindcss-mangle/config'
import type { InternalCacheOptions, InternalPatchOptions, PackageInfo, TailwindcssClassCache, TailwindcssPatcherOptions, TailwindcssRuntimeContext } from '../types'
import { createRequire } from 'node:module'
import process from 'node:process'
import fs from 'fs-extra'
import { getPackageInfoSync } from 'local-pkg'
import path from 'pathe'
import { getPatchOptions } from '../defaults'
import logger from '../logger'
import { isObject } from '../utils'
import { CacheManager, getCacheOptions } from './cache'
import { extractValidCandidates } from './candidates'
import { processTailwindcss } from './postcss'
import { internalPatch } from './runtime'

const require = createRequire(import.meta.url)
export class TailwindcssPatcher {
  public rawOptions: TailwindcssPatcherOptions
  public cacheOptions: InternalCacheOptions
  public patchOptions: InternalPatchOptions
  public patch: () => void
  public cacheManager: CacheManager
  public packageInfo: PackageInfo
  public majorVersion?: number

  constructor(options: TailwindcssPatcherOptions = {}) {
    this.rawOptions = options
    this.cacheOptions = getCacheOptions(options.cache)
    this.patchOptions = getPatchOptions(options.patch)

    this.cacheManager = new CacheManager(this.cacheOptions)

    const packageInfo = getPackageInfoSync('tailwindcss')

    if (!packageInfo) {
      throw new Error('tailwindcss not found')
    }

    if (packageInfo.version) {
      this.majorVersion = Number.parseInt(packageInfo.version[0])
    }
    this.packageInfo = packageInfo
    this.patch = () => {
      try {
        return internalPatch(this.packageInfo?.packageJsonPath, this.patchOptions)
      }
      catch (error) {
        logger.error(`patch tailwindcss failed: ${(<Error>error).message}`)
      }
    }
  }

  setCache(set: Set<string>) {
    if (this.cacheOptions.enable) {
      return this.cacheManager.write(set)
    }
  }

  getCache() {
    return this.cacheManager.read()
  }

  getContexts(): TailwindcssRuntimeContext[] {
    if (this.packageInfo) {
      const distPath = path.join(this.packageInfo.rootPath, 'lib')
      let injectFilePath: string | undefined
      if (this.majorVersion === 2) {
        injectFilePath = path.join(distPath, 'jit/index.js')
      }
      else if (this.majorVersion === 3) {
        injectFilePath = path.join(distPath, 'plugin.js')
        if (!fs.existsSync(injectFilePath)) {
          injectFilePath = path.join(distPath, 'index.js')
        }
      }
      if (injectFilePath) {
        const mo = require(injectFilePath)
        if (mo.contextRef) {
          return mo.contextRef.value
        }
      }
    }

    return []
  }

  getClassCaches(): TailwindcssClassCache[] {
    const contexts = this.getContexts()
    return contexts.filter(x => isObject(x)).map(x => x.classCache)
  }

  async getClassCacheSet(options?: PatchUserConfig): Promise<Set<string>> {
    const classSet = new Set<string>()
    const { output, tailwindcss } = options ?? {}
    if (this.majorVersion === 4) {
      const { v4 } = tailwindcss ?? {}
      if (Array.isArray(v4?.cssEntries)) {
        const results = (
          await Promise.all(
            v4
              .cssEntries
              .map(async (x) => {
                if (await fs.exists(x)) {
                  const css = await fs.readFile(x, 'utf8')
                  return css
                }
                return false
              }),
          )
        ).filter(x => x) as string[]
        for (const css of results) {
          const candidates = await extractValidCandidates({
            base: v4?.base,
            css,
            sources: v4?.sources?.map((x) => {
              return {
                base: x.base ?? v4?.base ?? process.cwd(),
                pattern: x.pattern,
              }
            }),
          })
          for (const candidate of candidates) {
            classSet.add(candidate)
          }
        }
      }
      else {
        const candidates = await extractValidCandidates({
          base: v4?.base,
          css: v4?.css,
          sources: v4?.sources?.map((x) => {
            return {
              base: x.base ?? v4?.base ?? process.cwd(),
              pattern: x.pattern,
            }
          }),
        })
        for (const candidate of candidates) {
          classSet.add(candidate)
        }
      }
    }
    else {
      const classCaches = this.getClassCaches()

      for (const classCacheMap of classCaches) {
        const keys = classCacheMap.keys()
        for (const key of keys) {
          const v = key.toString()
          if (output?.removeUniversalSelector && v === '*') {
            continue
          }
          classSet.add(v)
        }
      }
    }

    return classSet
  }

  /**
   * @description 在多个 tailwindcss 上下文时，这个方法将被执行多次，所以策略上应该使用 append
   */
  async getClassSet(options?: PatchUserConfig) {
    const { output, tailwindcss } = options ?? {}
    const cacheStrategy = this.cacheOptions.strategy ?? 'merge'
    const set = await this.getClassCacheSet({
      output,
      tailwindcss,
    })
    if (cacheStrategy === 'overwrite') {
      set.size > 0 && this.setCache(set)
    }
    else if (cacheStrategy === 'merge') {
      const cacheSet = this.getCache()
      if (cacheSet) {
        for (const x of cacheSet) {
          set.add(x)
        }
      }
      this.setCache(set)
    }

    return set
  }

  async extract(options?: PatchUserConfig) {
    const { output, tailwindcss } = options ?? {}
    if (output && tailwindcss) {
      const { filename, loose } = output

      if (this.majorVersion === 3) {
        await processTailwindcss({
          ...tailwindcss,
          majorVersion: this.majorVersion,
        })
      }

      const set = await this.getClassSet({
        output,
        tailwindcss,
      })
      if (filename) {
        const classList = [...set]
        await fs.outputJSON(filename, classList, {
          spaces: loose ? 2 : undefined,
        })
        return filename
      }
    }
  }

  extractValidCandidates = extractValidCandidates
}
