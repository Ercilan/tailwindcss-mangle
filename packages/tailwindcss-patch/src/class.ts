import { getClassCacheSet, getContexts, getTailwindcssEntry } from './exposeContext'
import type { CacheOptions, InternalCacheOptions, InternalPatchOptions, TailwindcssPatcherOptions } from './type'
import { writeCache, readCache } from './cache'
import { createPatch, getPatchOptions } from './patcher'

export function getCacheOptions(options?: CacheOptions | boolean) {
  let cache: InternalCacheOptions
  switch (typeof options) {
    case 'undefined': {
      cache = {
        enable: false
      }
      break
    }
    case 'boolean': {
      cache = {
        enable: options
      }
      break
    }
    case 'object': {
      cache = { ...options, enable: true }
      break
    }
  }
  return cache
}
export class TailwindcssPatcher {
  public rawOptions: TailwindcssPatcherOptions
  public cacheOptions: InternalCacheOptions
  public patchOptions: InternalPatchOptions
  public patch: () => void
  constructor(options: TailwindcssPatcherOptions = {}) {
    this.rawOptions = options
    this.cacheOptions = getCacheOptions(options.cache)
    this.patchOptions = getPatchOptions(options.patch)
    this.patch = createPatch(this.patchOptions)
  }

  getPkgEntry(basedir?: string) {
    return getTailwindcssEntry(basedir)
  }

  setCache(set: Set<string>) {
    if (this.cacheOptions.enable) {
      return writeCache(set, this.cacheOptions)
    }
  }

  getCache() {
    // if(this.cache.enable){
    return readCache(this.cacheOptions)
    // }
  }

  getClassSet(basedir?: string) {
    const set = getClassCacheSet(basedir)
    set.size && this.setCache(set)
    return set
  }

  getContexts(basedir?: string) {
    return getContexts(basedir)
  }
}
