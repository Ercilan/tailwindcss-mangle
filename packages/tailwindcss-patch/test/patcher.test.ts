import { getInstalledPkgJsonPath } from '../src/patcher'
import path from 'path'

describe('patcher', () => {
  it('getInstalledPkgJsonPath common options', () => {
    const pkgJsonPath = getInstalledPkgJsonPath()
    expect(pkgJsonPath).toBeTruthy()
    pkgJsonPath && expect(path.relative(process.cwd(), pkgJsonPath)).toMatchSnapshot()
  })

  it('getInstalledPkgJsonPath baseDir options', () => {
    const pkgJsonPath = getInstalledPkgJsonPath({
      basedir: path.resolve(__dirname, '../../../')
    })
    expect(pkgJsonPath).toBeTruthy()
  })
})
