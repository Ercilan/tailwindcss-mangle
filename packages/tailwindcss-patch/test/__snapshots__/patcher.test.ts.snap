// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`patcher getInstalledPkgJsonPath baseDir options 1`] = `"node_modules\\tailwindcss\\package.json"`;

exports[`patcher getInstalledPkgJsonPath common options 1`] = `"node_modules\\tailwindcss\\package.json"`;
